File: products/lunaos/lunaos-engine/packages/gateway/src/middleware.ts
/**
 * Middleware System for Claude Agent Platform
 *
 * Provides comprehensive middleware with:
 * - Authentication middleware
 * - Rate limiting middleware
 * - Request validation middleware
 * - CORS middleware
 * - Logging middleware
 * - Error handling middleware
 */

import { Request, Response, NextFunction } from 'express';
import { RequestValidator } from 'express-validator';
import { RedisRateLimit } from 'express-rate-limit';
import { RedisCache } from '@claude-agent/cache';
import { AuthenticationService } from './auth';
import {
  MiddlewareOptions,
  AuthenticatedRequest,
  RateLimitInfo,
  GatewayConfig,
  ErrorResponse,
  SuccessResponse
} from './interfaces';

export class MiddlewareManager {
  private authService: AuthenticationService;
  private cache: RedisCache;
  private config: GatewayConfig;

  constructor(
    authService: AuthenticationService,
    cache: RedisCache,
    config: GatewayConfig
  ) {
    this.authService = authService;
    this.cache = cache;
    this.config = config;
  }

  /**
   * Create authentication middleware
   */
  createAuthMiddleware(options: MiddlewareOptions['authentication'] = {}) {
    return this.authService.createAuthMiddleware(options);
  }

  /**
   * Create rate limiting middleware
   */
  createRateLimitMiddleware(options: MiddlewareOptions['rateLimit'] = {}) {
    const keyGenerator = options.keyGenerator || this.defaultKeyGenerator;

    const rateLimiter = new RedisRateLimit({
      store: {
        type: 'redis',
        client: this.cache.getClient() as any,
        prefix: 'rate-limit:',
      },
      keyGenerator: (req: AuthenticatedRequest) => {
        // Use API key or user ID for rate limiting
        if (req.apiKey) {
          return `api_key:${req.apiKey.id}`;
        }
        if (req.user) {
          return `user:${req.user.id}`;
        }
        return `ip:${req.ip}`;
      },
      windowMs: options.windowMs || this.config.rateLimiting.windowMs,
      max: options.maxRequests || this.config.rateLimiting.maxRequests,
      skipSuccessfulRequests: options.skipSuccessfulRequests !== false,
      skipFailedRequests: options.skipFailedRequests !== false,
      onLimitReached: async (req: AuthenticatedRequest, res: Response) => {
        res.status(429).json({
          error: {
            code: 'RATE_LIMIT_EXCEEDED',
            message: 'Too many requests, please try again later',
            requestId: req.id || require('uuid').v4(),
            timestamp: new Date(),
            details: {
              limit: rateLimiter.store.options.max,
              current: rateLimiter.store.options.max, // Would need to track current
              resetTime: new Date(Date.now() + rateLimiter.store.options.windowMs),
            },
          },
        });
      },
    } as any);

    return rateLimiter;
  }

  /**
   * Create CORS middleware
   */
  createCORSMiddleware(options: MiddlewareOptions['cors'] = {}) {
    const corsConfig = {
      origin: options.origin || this.config.cors.origin,
      credentials: options.credentials !== false,
      methods: options.methods || ['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'OPTIONS'],
      allowedHeaders: options.headers || ['Content-Type', 'Authorization', 'X-Requested-With', 'X-API-Key'],
      exposedHeaders: ['X-RateLimit-Limit', 'X-RateLimit-Remaining', 'X-RateLimit-Reset'],
      maxAge: 86400, // 24 hours
      optionsSuccessStatus: 200,
    };

    return require('cors')(corsConfig);
  }

  /**
   * Create request validation middleware
   */
  createValidationMiddleware(schema: MiddlewareOptions['validation']) {
    const validation = new RequestValidator(schema || {});

    return async (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
      try {
        // Validate body
        if (schema.body) {
          await req.body(schema.body);
        }

        // Validate query parameters
        if (schema.query) {
          await req.query(schema.query);
        }

        // Validate URL parameters
        if (schema.params) {
          await req.params(schema.params);
        }

        // Validate headers
        if (schema.headers) {
          await req.headers(schema.headers);
        }

        next();
      } catch (error) {
        const errorDetails = this.formatValidationError(error);

        res.status(400).json({
          error: {
            code: 'VALIDATION_ERROR',
            message: 'Request validation failed',
            details: errorDetails,
            requestId: req.id || require('uuid').v4(),
            timestamp: new Date(),
          },
        });
      }
    };
  }

  /**
   * Create response middleware
   */
  createResponseMiddleware() {
    return (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
      // Add rate limit headers if available
      if ((req as any).rateLimit) {
        const rateLimit = (req as any).rateLimit as RateLimitInfo;
        res.set('X-RateLimit-Limit', rateLimit.limit.toString());
        res.set('X-RateLimit-Remaining', rateLimit.remaining.toString());
        res.set('X-RateLimit-Reset', rateLimit.resetTime.toISOString());
      }

      // Add security headers
      res.set('X-Content-Type-Options', 'nosniff');
      res.set('X-Frame-Options', 'DENY');
      res.set('X-XSS-Protection', '1; mode=block');
      res.set('Referrer-Policy', 'strict-origin-when-cross-origin');
      res.set('Content-Security-Policy', "default-src 'self'");

      // Add response ID
      res.set('X-Request-ID', req.id || require('uuid').v4());

      // Add API version
      res.set('X-API-Version', '1.0.0');

      // Format response
      const originalJson = res.json;
      res.json = function(body: any) {
        const response = this.formatResponse(body);
        originalJson.call(this, response);
      }.bind(res);

      next();
    };
  }

  /**
   * Create logging middleware
   */
  createLoggingMiddleware() {
    return (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
      const startTime = Date.now();

      // Log request
      console.log(JSON.stringify({
        type: 'request',
        timestamp: new Date().toISOString(),
        method: req.method,
        url: req.originalUrl,
        userAgent: req.get('User-Agent'),
        ip: req.ip,
        userId: req.user?.id,
        apiKeyId: req.apiKey?.id,
        requestId: req.id,
      }));

      // Override res.end to log response
      const originalEnd = res.end;
      res.end = function(chunk?: any) {
        const endTime = Date.now();
        const responseTime = endTime - startTime;

        console.log(JSON.stringify({
          type: 'response',
          timestamp: new Date().toISOString(),
          method: req.method,
          url: req.originalUrl,
          statusCode: res.statusCode,
          responseTime,
          userId: req.user?.id,
          apiKeyId: req.apiKey?.id,
          requestId: req.id,
        }));

        originalEnd.call(this, chunk);
      }.bind(res);

      next();
    };
  }

  /**
   * Create error handling middleware
   */
  createErrorHandlingMiddleware() {
    return (error: Error, req: AuthenticatedRequest, res: Response, next: NextFunction) => {
      console.error(JSON.stringify({
        type: 'error',
        timestamp: new Date().toISOString(),
        method: req.method,
        url: req.originalUrl,
        statusCode: res.statusCode,
        error: {
          name: error.name,
          message: error.message,
          stack: error.stack,
        },
        userId: req.user?.id,
        apiKeyId: req.apiKey?.id,
        requestId: req.id,
      }));

      // Don't respond if headers already sent
      if (res.headersSent) {
        return next(error);
      }

      // Handle validation errors
      if (error.name === 'ValidationError') {
        res.status(400).json({
          error: {
            code: 'VALIDATION_ERROR',
            message: 'Request validation failed',
            details: this.formatValidationError(error),
            requestId: req.id || require('uuid').v4(),
            timestamp: new Date(),
          },
        });
        return;
      }

      // Handle JWT errors
      if (error.name === 'JsonWebTokenError') {
        res.status(401).json({
          error: {
            code: 'INVALID_TOKEN',
            message: 'Invalid or expired token',
            requestId: req.id || require('uuid').v4(),
            timestamp: new Date(),
          },
        });
        return;
      }

      // Handle rate limit errors
      if (error.name === 'RateLimitExceeded') {
        res.status(429).json({
          error: {
            code: 'RATE_LIMIT_EXCEEDED',
            message: 'Rate limit exceeded',
            requestId: req.id || require('uuid').v4(),
            timestamp: new Date(),
          },
        });
        return;
      }

      // Handle API key errors
      if (error.name === 'UnauthorizedError') {
        res.status(401).json({
          error: {
            code: 'INVALID_API_KEY',
            message: 'Invalid API key',
            requestId: req.id || require('uuid').v4(),
            timestamp: new Date(),
          },
        });
        return;
      }

      // Handle Prisma errors
      if (error.name === 'PrismaClientKnownRequestError') {
        res.status(503).json({
          error: {
            code: 'DATABASE_ERROR',
            message: 'Database operation failed',
            requestId: req.id || require('uuid').v4(),
            timestamp: new Date(),
          },
        });
        return;
      }

      // Handle generic errors
      const statusCode = this.getHttpStatus(error);
      res.status(statusCode).json({
        error: {
          code: 'INTERNAL_SERVER_ERROR',
          message: process.env.NODE_ENV === 'production' ? 'Internal server error' : error.message,
          requestId: req.id || require('uuid').v4(),
          timestamp: new Date(),
          stack: process.env.NODE_ENV === 'development' ? error.stack : undefined,
        },
      });
    };
  }

  /**
   * Create cache middleware
   */
  createCacheMiddleware(options: MiddlewareOptions['cache'] = {}) {
    const keyGenerator = options.keyGenerator || this.defaultCacheKeyGenerator;
    const ttl = options.ttl || 300000; // 5 minutes

    return async (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
      const cacheKey = keyGenerator(req);

      try {
        const cachedResult = await this.cache.get<{ response: any }>(cacheKey);

        if (cachedResult && cachedResult.hit) {
          return res.json(cachedResult.value.response);
        }

        // Override res.json to cache successful GET requests
        const originalJson = res.json;
        res.json = function(body: any) {
          if (req.method === 'GET' && res.statusCode < 400) {
            this.cache.set(cacheKey, { response: body }, { ttl });
          }
          originalJson.call(this, body);
        }.bind(res);

        next();
      } catch (error) {
        console.error('Cache middleware error:', error);
        next();
      }
    };
  }

  /**
   * Default key generator for rate limiting
   */
  private defaultKeyGenerator(req: AuthenticatedRequest): string {
    if (req.apiKey) {
      return `api_key:${req.apiKey.id}`;
    }
    if (req.user) {
      return `user:${req.user.id}`;
    }
    return `ip:${req.ip}`;
  }

  /**
   * Default cache key generator
   */
  private defaultCacheKeyGenerator(req: AuthenticatedRequest): string {
    const userId = req.user?.id || 'anonymous';
    const path = req.path;
    const query = JSON.stringify(req.query);
    return `cache:${userId}:${path}:${query}`;
  }

  /**
   * Format response consistently
   */
  private formatResponse(body: any): SuccessResponse<any> | ErrorResponse {
    if (body && typeof body === 'object' && 'error' in body) {
      return body as ErrorResponse;
    }

    return {
      data: body,
      meta: {
        requestId: require('uuid').v4(),
        timestamp: new Date(),
        version: '1.0.0',
      },
    };
  }

  /**
   * Get HTTP status code from error
   */
  private getHttpStatus(error: Error): number {
    if (error.message.includes('not found')) {
      return 404;
    }
    if (error.message.includes('unauthorized') || error.message.includes('forbidden')) {
      return 403;
    }
    if (error.message.includes('validation')) {
      return 400;
    }
    if (error.message.includes('timeout')) {
      return 504;
    }
    return 500;
  }

  /**
   * Format validation error details
   */
  private formatValidationError(error: any): any {
    if (error.details) {
      return error.details;
    }

    // Handle express-validator specific format
    if (error.array) {
      return {
        errors: error.array,
        message: error.message,
      };
    }

    return {
      message: error.message,
    };
  }
}

File: products/sdlc-cc/platform/services/gateway/internal/infrastructure/middleware/middleware.go
//go:build ignore

package middleware

import (
	"compress/gzip"
	"context"
	"fmt"
	"net/http"
	"runtime/debug"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/go-chi/cors"
	"github.com/go-chi/render"
	"github.com/sirupsen/logrus"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/trace"

	"github.com/sdlc-ai/platform/services/gateway/internal/infrastructure/config"
)

// LoggingMiddleware creates a structured logging middleware
func LoggingMiddleware() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()

			// Create response writer wrapper to capture status code
			wrapped := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}

			// Log request
			logger := logrus.WithFields(logrus.Fields{
				"method":      r.Method,
				"path":        r.URL.Path,
				"query":       r.URL.RawQuery,
				"remote_addr": r.RemoteAddr,
				"user_agent":  r.UserAgent(),
				"request_id":  r.Header.Get("X-Request-ID"),
			})

			logger.Info("Request started")

			// Process request
			next.ServeHTTP(wrapped, r)

			// Log response
			duration := time.Since(start)
			logger.WithFields(logrus.Fields{
				"status_code": wrapped.statusCode,
				"duration_ms": duration.Milliseconds(),
				"size_bytes":  wrapped.size,
			}).Info("Request completed")
		})
	}
}

// TracingMiddleware creates an OpenTelemetry tracing middleware
func TracingMiddleware() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Extract trace context from incoming headers
			propagator := propagation.TraceContext{}
			ctx := propagator.Extract(r.Context(), propagation.HeaderCarrier(r.Header))

			// Start span
			tracer := otel.Tracer("gateway")
			ctx, span := tracer.Start(ctx, fmt.Sprintf("%s %s", r.Method, r.URL.Path),
				trace.WithAttributes(
					attribute.String("http.method", r.Method),
					attribute.String("http.url", r.URL.String()),
					attribute.String("http.host", r.Host),
					attribute.String("http.scheme", "http"),
					attribute.String("http.user_agent", r.UserAgent()),
					attribute.String("http.remote_addr", r.RemoteAddr),
				),
			)
			defer span.End()

			// Continue with trace context
			r = r.WithContext(ctx)

			// Wrap response writer to capture status code
			wrapped := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}

			// Process request
			next.ServeHTTP(wrapped, r)

			// Add response attributes to span
			span.SetAttributes(
				attribute.Int("http.status_code", wrapped.statusCode),
				attribute.Int("http.response_size", wrapped.size),
			)

			// Mark span as error if status code indicates error
			if wrapped.statusCode >= 400 {
				span.SetAttributes(attribute.String("error", fmt.Sprintf("HTTP %d", wrapped.statusCode)))
				if wrapped.statusCode >= 500 {
					span.RecordError(fmt.Errorf("server error: %d", wrapped.statusCode))
				}
			}
		})
	}
}

// CORSMiddleware creates CORS middleware with configurable options
func CORSMiddleware(cfg *config.Config) func(http.Handler) http.Handler {
	corsOptions := cors.Options{
		AllowedOrigins:   cfg.CORS.AllowedOrigins,
		AllowedMethods:   []string{"GET", "POST", "PUT", "DELETE", "OPTIONS", "PATCH"},
		AllowedHeaders:   []string{"Accept", "Authorization", "Content-Type", "X-CSRF-Token", "X-Request-ID", "X-Tenant-ID"},
		ExposedHeaders:   []string{"X-Total-Count", "X-Rate-Limit-Limit", "X-Rate-Limit-Remaining", "X-Rate-Limit-Reset"},
		AllowCredentials: true,
		MaxAge:           300, // 5 minutes
		Debug:            cfg.Environment == "development",
	}

	return cors.Handler(corsOptions)
}

// SecurityHeadersMiddleware adds security headers to responses
func SecurityHeadersMiddleware() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Security headers
			w.Header().Set("X-Content-Type-Options", "nosniff")
			w.Header().Set("X-Frame-Options", "DENY")
			w.Header().Set("X-XSS-Protection", "1; mode=block")
			w.Header().Set("Referrer-Policy", "strict-origin-when-cross-origin")
			w.Header().Set("Permissions-Policy", "camera=(), microphone=(), geolocation=()")
			w.Header().Set("Cross-Origin-Embedder-Policy", "require-corp")
			w.Header().Set("Cross-Origin-Opener-Policy", "same-origin")
			w.Header().Set("Cross-Origin-Resource-Policy", "same-origin")

			// HSTS header — always set behind reverse proxy (X-Forwarded-Proto)
			if r.TLS != nil || r.Header.Get("X-Forwarded-Proto") == "https" {
				w.Header().Set("Strict-Transport-Security", "max-age=63072000; includeSubDomains; preload")
			}

			// Prevent caching of API responses containing sensitive data
			w.Header().Set("Cache-Control", "no-store, no-cache, must-revalidate, private")
			w.Header().Set("Pragma", "no-cache")

			// Content Security Policy — strict for API endpoints (no inline scripts)
			csp := "default-src 'none'; " +
				"frame-ancestors 'none'; " +
				"base-uri 'none'; " +
				"form-action 'none'; " +
				"sandbox"
			w.Header().Set("Content-Security-Policy", csp)

			next.ServeHTTP(w, r)
		})
	}
}

// CompressionMiddleware adds gzip compression
func CompressionMiddleware() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Check if client accepts gzip
			if !strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") {
				next.ServeHTTP(w, r)
				return
			}

			// Don't compress already compressed content
			if strings.Contains(r.Header.Get("Content-Encoding"), "gzip") {
				next.ServeHTTP(w, r)
				return
			}

			// Create gzip writer
			gzipWriter := &gzipResponseWriter{
				ResponseWriter: w,
				gzipWriter:     gzip.NewWriter(w),
			}
			defer gzipWriter.gzipWriter.Close()

			// Set appropriate headers
			w.Header().Set("Content-Encoding", "gzip")
			w.Header().Set("Vary", "Accept-Encoding")

			next.ServeHTTP(gzipWriter, r)
		})
	}
}

// RateLimitMiddleware creates a rate limiting middleware
func RateLimitMiddleware(cfg *config.Config) func(http.Handler) http.Handler {
	// In-memory rate limiter (in production, use Redis or similar)
	limiter := &rateLimiter{
		clients: make(map[string]*clientLimiter),
		mutex:   &sync.RWMutex{},
		limit:   cfg.RateLimit.Requests,
		window:  cfg.RateLimit.Window,
	}

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Get client identifier (IP or user ID)
			clientID := getClientIdentifier(r)

			// Check rate limit
			if !limiter.allow(clientID) {
				// Set rate limit headers
				w.Header().Set("X-RateLimit-Limit", strconv.Itoa(cfg.RateLimit.Requests))
				w.Header().Set("X-RateLimit-Remaining", "0")
				w.Header().Set("X-RateLimit-Reset", strconv.Itoa(int(time.Now().Add(cfg.RateLimit.Window).Unix())))

				render.Status(r, http.StatusTooManyRequests)
				render.JSON(w, r, map[string]interface{}{
					"error": map[string]interface{}{
						"code":    "RATE_LIMIT_EXCEEDED",
						"message": "Rate limit exceeded",
					},
					"meta": map[string]interface{}{
						"timestamp": time.Now().UTC().Format(time.RFC3339),
					},
				})
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}

// CircuitBreakerMiddleware creates a circuit breaker middleware
func CircuitBreaker(cfg *config.Config) func(http.Handler) http.Handler {
	breaker := &circuitBreaker{
		maxFailures:  cfg.CircuitBreaker.MaxFailures,
		resetTimeout: cfg.CircuitBreaker.ResetTimeout,
		state:        circuitClosed,
		lastFailure:  time.Now(),
	}

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if !breaker.allow() {
				render.Status(r, http.StatusServiceUnavailable)
				render.JSON(w, r, map[string]interface{}{
					"error": map[string]interface{}{
						"code":    "SERVICE_UNAVAILABLE",
						"message": "Service temporarily unavailable",
					},
					"meta": map[string]interface{}{
						"timestamp": time.Now().UTC().Format(time.RFC3339),
					},
				})
				return
			}

			// Wrap response writer to capture failures
			wrapped := &circuitBreakerResponseWriter{
				ResponseWriter: w,
				breaker:        breaker,
				statusCode:     http.StatusOK,
			}

			next.ServeHTTP(wrapped, r)
		})
	}
}

// AuthenticationMiddleware creates JWT authentication middleware
func AuthenticationMiddleware(cfg *config.Config) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Skip authentication for health checks and metrics
			if isPublicPath(r.URL.Path) {
				next.ServeHTTP(w, r)
				return
			}

			// Extract JWT token from Authorization header
			authHeader := r.Header.Get("Authorization")
			if authHeader == "" {
				render.Status(r, http.StatusUnauthorized)
				render.JSON(w, r, map[string]interface{}{
					"error": map[string]interface{}{
						"code":    "MISSING_TOKEN",
						"message": "Authorization header required",
					},
				})
				return
			}

			// Validate token format
			tokenParts := strings.Split(authHeader, " ")
			if len(tokenParts) != 2 || tokenParts[0] != "Bearer" {
				render.Status(r, http.StatusUnauthorized)
				render.JSON(w, r, map[string]interface{}{
					"error": map[string]interface{}{
						"code":    "INVALID_TOKEN_FORMAT",
						"message": "Invalid authorization header format",
					},
				})
				return
			}

			// TODO: Implement actual JWT validation
			// For now, we'll just pass through
			next.ServeHTTP(w, r)
		})
	}
}

// AuthorizationMiddleware creates OPA policy evaluation middleware
func AuthorizationMiddleware(policyEngine interface{}) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Skip authorization for health checks and metrics
			if isPublicPath(r.URL.Path) {
				next.ServeHTTP(w, r)
				return
			}

			// TODO: Implement actual OPA policy evaluation
			// For now, we'll just pass through
			next.ServeHTTP(w, r)
		})
	}
}

// RecoveryMiddleware creates a recovery middleware that catches panics
func RecoveryMiddleware() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			defer func() {
				if err := recover(); err != nil {
					// Log the panic
					logrus.WithFields(logrus.Fields{
						"error":       err,
						"stack":       string(debug.Stack()),
						"method":      r.Method,
						"path":        r.URL.Path,
						"remote_addr": r.RemoteAddr,
						"user_agent":  r.UserAgent(),
					}).Error("Request panic recovered")

					// Return error response
					render.Status(r, http.StatusInternalServerError)
					render.JSON(w, r, map[string]interface{}{
						"error": map[string]interface{}{
							"code":    "INTERNAL_SERVER_ERROR",
							"message": "Internal server error",
						},
						"meta": map[string]interface{}{
							"timestamp": time.Now().UTC().Format(time.RFC3339),
						},
					})
				}
			}()

			next.ServeHTTP(w, r)
		})
	}
}

// RequestTimeoutMiddleware adds timeout to requests
func RequestTimeoutMiddleware(timeout time.Duration) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx, cancel := context.WithTimeout(r.Context(), timeout)
			defer cancel()

			r = r.WithContext(ctx)
			next.ServeHTTP(w, r)
		})
	}
}

// Helper types and functions

type responseWriter struct {
	http.ResponseWriter
	statusCode int
	size       int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

func (rw *responseWriter) Write(b []byte) (int, error) {
	n, err := rw.ResponseWriter.Write(b)
	rw.size += n
	return n, err
}

type gzipResponseWriter struct {
	http.ResponseWriter
	gzipWriter *gzip.Writer
}

func (grw *gzipResponseWriter) Write(b []byte) (int, error) {
	return grw.gzipWriter.Write(b)
}

type circuitBreakerState int

const (
	circuitClosed circuitBreakerState = iota
	circuitOpen
	circuitHalfOpen
)

type circuitBreaker struct {
	maxFailures  int
	resetTimeout time.Duration
	state        circuitBreakerState
	failures     int
	lastFailure  time.Time
	mutex        sync.RWMutex
}

func (cb *circuitBreaker) allow() bool {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	switch cb.state {
	case circuitClosed:
		return true
	case circuitOpen:
		if time.Since(cb.lastFailure) > cb.resetTimeout {
			cb.state = circuitHalfOpen
			cb.failures = 0
			return true
		}
		return false
	case circuitHalfOpen:
		return true
	default:
		return false
	}
}

func (cb *circuitBreaker) recordSuccess() {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	cb.failures = 0
	if cb.state == circuitHalfOpen {
		cb.state = circuitClosed
	}
}

func (cb *circuitBreaker) recordFailure() {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	cb.failures++
	cb.lastFailure = time.Now()

	if cb.failures >= cb.maxFailures {
		cb.state = circuitOpen
	}
}

type circuitBreakerResponseWriter struct {
	http.ResponseWriter
	breaker    *circuitBreaker
	statusCode int
}

func (cbw *circuitBreakerResponseWriter) WriteHeader(code int) {
	cbw.statusCode = code
	if code >= 500 {
		cbw.breaker.recordFailure()
	} else {
		cbw.breaker.recordSuccess()
	}
	cbw.ResponseWriter.WriteHeader(code)
}

type rateLimiter struct {
	clients map[string]*clientLimiter
	mutex   *sync.RWMutex
	limit   int
	window  time.Duration
}

type clientLimiter struct {
	tokens     int
	lastRefill time.Time
	mutex      sync.Mutex
}

func (rl *rateLimiter) allow(clientID string) bool {
	rl.mutex.RLock()
	client, exists := rl.clients[clientID]
	rl.mutex.RUnlock()

	if !exists {
		rl.mutex.Lock()
		client = &clientLimiter{
			tokens:     rl.limit,
			lastRefill: time.Now(),
		}
		rl.clients[clientID] = client
		rl.mutex.Unlock()
	}

	client.mutex.Lock()
	defer client.mutex.Unlock()

	// Refill tokens based on time elapsed
	now := time.Now()
	elapsed := now.Sub(client.lastRefill)
	tokensToAdd := int(elapsed.Seconds()) * (rl.limit / int(rl.window.Seconds()))

	if tokensToAdd > 0 {
		client.tokens = min(rl.limit, client.tokens+tokensToAdd)
		client.lastRefill = now
	}

	if client.tokens > 0 {
		client.tokens--
		return true
	}

	return false
}

func getClientIdentifier(r *http.Request) string {
	// Try to get user ID from context (if authenticated)
	// For now, use IP address as fallback
	return r.RemoteAddr
}

func isPublicPath(path string) bool {
	publicPaths := []string{
		"/health",
		"/healthz",
		"/ready",
		"/readyz",
		"/live",
		"/livez",
		"/metrics",
		"/version",
	}

	for _, publicPath := range publicPaths {
		if strings.HasPrefix(path, publicPath) {
			return true
		}
	}
	return false
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
